quickwit/quickwit-search/src/leaf.rs

```diff
@@ -1000,134 +995,236 @@ impl CanSplitDoBetter {
+            let Some(split_start) = split.timestamp_start else {
+                return false;
+            };
+            if split_start < start {
+                return false;
+            }
+        }
+        if let Some(end) = search_request.end_timestamp {
+            let Some(split_end) = split.timestamp_end else {
+                return false;
+            };
+            if split_end >= end {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn to_splits_with_request(
+        splits: Vec<SplitIdAndFooterOffsets>,
+        request: Arc<SearchRequest>,
+    ) -> Vec<(SplitIdAndFooterOffsets, SearchRequest)> {
+        splits
+            .into_iter()
+            .map(|split| (split, (*request).clone()))
+            .collect::<Vec<_>>()
+    }
+
+    /// Calculate the number of splits which are guaranteed to deliver enough documents.
     ///
-    /// The leaf search code contains some logic that makes it possible to skip entire splits
-    /// when we are confident they won't make it into top K.
-    /// To make this optimization as potent as possible, we sort the splits so that the first splits
-    /// are the most likely to fill our Top K.
-    /// In the future, as split get more metadata per column, we may be able to do this more than
-    /// just for timestamp and "unsorted" request.
-    fn optimize_split_order(&self, splits: &mut [SplitIdAndFooterOffsets]) {
-        match self {
-            CanSplitDoBetter::SplitIdHigher(_) => {
-                splits.sort_unstable_by(|a, b| b.split_id.cmp(&a.split_id))
+    /// If there's a time range and not enough splits contain at least the number of requested
+    /// documents, return None.
+    fn get_min_required_splits(
+        splits: &[SplitIdAndFooterOffsets],
+        request: &SearchRequest,
+    ) -> Option<usize> {
+        let num_requested_docs = request.start_offset + request.max_hits;
+
+        let mut min_required_splits = 0;
+        let mut partial_sum = 0;
+
+        for split in splits.iter() {
+            if !Self::is_split_contained_in_search_time_range(split, &request) {
+                continue;
             }
-            CanSplitDoBetter::SplitTimestampHigher(_)
-            | CanSplitDoBetter::FindTraceIdsAggregation(_) => {
-                splits.sort_unstable_by_key(|split| std::cmp::Reverse(split.timestamp_end()))
+
+            partial_sum += split.num_docs;
+
+            if partial_sum >= num_requested_docs {
+                return Some(min_required_splits + 1);
             }
-            CanSplitDoBetter::SplitTimestampLower(_) => {
-                splits.sort_unstable_by_key(|split| split.timestamp_start())
+
+            min_required_splits += 1;
+        }
+
+        None
+    }
+
+    fn optimize_split_id_higher(
+        &self,
+        request: Arc<SearchRequest>,
+        mut splits: Vec<SplitIdAndFooterOffsets>,
+    ) -> Result<Vec<(SplitIdAndFooterOffsets, SearchRequest)>, SearchError> {
```

Comment:
we could actually call `to_splits_with_request` before (inside `optimize`) and simplify this signature to something like:
```suggestion
    fn optimize_split_id_higher(
       &mut [(SplitIdAndFooterOffsets, SearchRequest)]
    ) -> Result<(), SearchError> {
```
This would prevent quite a lot of repetition and is also more explicit because it shows that we don't drop splits (note that `&self` is also dropped because not used)
(applies to all `optimzie_split_xxx`)